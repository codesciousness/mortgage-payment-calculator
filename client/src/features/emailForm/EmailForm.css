.EmailForm {
    margin-top: 1.5rem;
    border-radius: 4px;
    background-color: rgba(135, 206, 235, 0.15); /* skyblue */
    padding: 0.5rem 0;
}

.EmailForm__header {
    margin: 0.5rem 1rem;
}

.EmailForm__container {
    display: flex;
    align-items: center;
}

@media only screen and (max-width: 1200px) and (min-width: 1001px) {

    .EmailForm__header {
        text-align: center;
        margin: 1rem;
    }

    .EmailForm__container {
        flex-direction: column;
    }
}

@media only screen and (max-width: 700px) {

    .EmailForm__header {
        text-align: center;
        margin: 1rem;
    }

    .EmailForm__container {
        flex-direction: column;
    }
}