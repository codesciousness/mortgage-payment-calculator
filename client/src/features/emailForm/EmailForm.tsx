import React, { useEffect } from 'react';
import Button from '../../components/button/Button';
import TextInput from '../../components/textInput/TextInput';
import Alert from '../../components/alert/Alert';
import './EmailForm.css';
import { selectName, selectEmail, setName, setEmail, selectHomePrice, selectDownPayment,
    selectLoanTerm, selectInterestRate, selectPropertyTax, selectHomeInsurance, selectPMI,
    selectHOAFees, selectStartDate, selectMortgagePayment, selectMonthlyPayment, selectLoanAmount, 
    selectTotalInterest, selectLoanCost, selectPayoffDate, selectAmortizationSchedule, saveLoan,
    selectSavingLoan, selectSaveLoanSuccess, selectSaveLoanError, clearStatusUpdates } from '../loansSlice';
import { useAppSelector, useAppDispatch } from '../../app/hooks';
import { useWindowSize } from '../../hooks/use-window-size';

type handleChangeProps = {
    target: HTMLInputElement;
}

const EmailForm = (): JSX.Element => {
    const name = useAppSelector(selectName);
    const email = useAppSelector(selectEmail);
    const homePrice = useAppSelector(selectHomePrice);
    const downPayment = useAppSelector(selectDownPayment);
    const loanTerm = useAppSelector(selectLoanTerm);
    const interestRate = useAppSelector(selectInterestRate);
    const startDate = useAppSelector(selectStartDate);
    const propertyTax = useAppSelector(selectPropertyTax);
    const homeInsurance = useAppSelector(selectHomeInsurance);
    const privateMortgageInsurance = useAppSelector(selectPMI);
    const hoaFees = useAppSelector(selectHOAFees);
    const mortgagePayment = useAppSelector(selectMortgagePayment);
    const monthlyPayment = useAppSelector(selectMonthlyPayment);
    const loanAmount = useAppSelector(selectLoanAmount);
    const loanCost = useAppSelector(selectLoanCost);
    const totalInterest = useAppSelector(selectTotalInterest);
    const payoffDate = useAppSelector(selectPayoffDate);
    const amortizationSchedule = useAppSelector(selectAmortizationSchedule);
    const savingLoan = useAppSelector(selectSavingLoan);
    const saveLoanSuccess = useAppSelector(selectSaveLoanSuccess);
    const saveLoanError = useAppSelector(selectSaveLoanError);
    const dispatch = useAppDispatch();
    const size = useWindowSize();
    const width = size.width && size.width > 350 ? 300 : '90%';

    const loan = {
        name,
        email,
        homePrice,
        downPayment,
        loanTerm,
        interestRate,
        startDate,
        propertyTax,
        homeInsurance,
        privateMortgageInsurance,
        hoaFees,
        mortgagePayment,
        monthlyPayment,
        loanAmount,
        loanCost,
        totalInterest,
        payoffDate,
        amortizationSchedule
    };

    const handleChange = ({ target }: handleChangeProps) => {
        const { id, value } = target;

        if (id === 'Name') dispatch(setName(value));
        else if (id === 'Email') dispatch(setEmail(value));
    };

    const handleClick = () => dispatch(saveLoan(loan));

    useEffect(() => {
        if (saveLoanSuccess) {
            dispatch(setName(''));
            dispatch(setEmail(''));
        }
    }, [saveLoanSuccess, dispatch]);

    return (
        <section id='EmailForm' className='EmailForm'>
            <h3 className='EmailForm__header'>Email your mortgage payment summary!</h3>
            <div className='EmailForm__container'>
                <TextInput name='Name' value={name} width={width} onChange={handleChange}/>
                <TextInput name='Email' value={email} width={width} onChange={handleChange}/>
                <Button name='Send' loading={savingLoan} onClick={handleClick}/>
            </div>
            {saveLoanError && <Alert severity='error' msg={saveLoanError} onClose={() => dispatch(clearStatusUpdates())}/>}
            {saveLoanSuccess && <Alert severity='success' msg='Loan data submitted successfully!' onClose={() => dispatch(clearStatusUpdates())}/>}
        </section>
    );
};

export default EmailForm;