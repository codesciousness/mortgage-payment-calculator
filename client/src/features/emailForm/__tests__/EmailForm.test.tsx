import React from 'react';
import { Provider } from 'react-redux';
import { render, screen, fireEvent } from '@testing-library/react';
import { create } from 'react-test-renderer';
import { store } from '../../../app/store';
import { rest } from 'msw';
import { setupServer } from 'msw/node';
import EmailForm from '../EmailForm';

const server = setupServer(
    rest.post('/loans', (req, res, ctx) => {
        return res(ctx.json({}));
    }),
);
  
beforeAll(() => server.listen());
afterEach(() => server.resetHandlers());
afterAll(() => server.close());

describe('<EmailForm />', () => {
    test('EmailForm component is defined', () => {
        expect(<EmailForm/>).toBeDefined();
    });
    test('component matches the snapshot', () => {
        const component = create(
            <Provider store={store}>
                <EmailForm/>
            </Provider>
        );
        expect(component.toJSON()).toMatchSnapshot();
    });
    test('renders correctly with initial state from Redux store', () => {
        render(
            <Provider store={store}>
                <EmailForm/>
            </Provider>
        );
        expect(screen.getByText('Email your mortgage payment summary!')).toBeInTheDocument();
        expect(screen.getByPlaceholderText('Enter your name')).toBeInTheDocument();
        expect(screen.getByPlaceholderText('Enter your email')).toBeInTheDocument();
    });
    test('displays a success alert when button is clicked and name and email are filled correctly', async () => {
        render(
            <Provider store={store}>
                <EmailForm/>
            </Provider>
        );
        expect(screen.queryByText('Loan data submitted successfully!')).not.toBeInTheDocument();
        fireEvent.change(screen.getByPlaceholderText('Enter your name'), {target: {value: 'Bob Test'}});
        fireEvent.change(screen.getByPlaceholderText('Enter your email'), {target: {value: 'bob.test@company.com'}});
        fireEvent.click(screen.getByText('Send'));
        await screen.findByText('Loan data submitted successfully!');
    });
    test('displays an error alert when button is clicked and name and email are blank', async () => {
        server.use(
            rest.post('/loans', (req, res, ctx) => {
                return res(
                    ctx.status(400),
                    ctx.json('Please provide a name and email address.')
                );
            }),
        );
        render(
            <Provider store={store}>
                <EmailForm/>
            </Provider>
        );
        expect(screen.queryByText('Please provide a name and email address.')).not.toBeInTheDocument();
        fireEvent.click(screen.getByText('Send'));
        await screen.findByText('Please provide a name and email address.');
    });
});