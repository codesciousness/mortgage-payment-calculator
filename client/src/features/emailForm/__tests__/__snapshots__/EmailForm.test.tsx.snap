// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EmailForm /> component matches the snapshot 1`] = `
<section
  className="EmailForm"
  id="EmailForm"
>
  <h3
    className="EmailForm__header"
  >
    Email your mortgage payment summary!
  </h3>
  <div
    className="EmailForm__container"
  >
    <div
      className="MuiFormControl-root MuiTextField-root css-1cs2p1o-MuiFormControl-root-MuiTextField-root"
    >
      <label
        className="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-1pysi21-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink={false}
        htmlFor="Name"
        id="Name-label"
      >
        Name
      </label>
      <div
        className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          autoFocus={false}
          className="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
          disabled={false}
          id="Name"
          name="Name"
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder="Enter your name"
          required={false}
          type="text"
          value=""
        />
        <fieldset
          aria-hidden={true}
          className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            className="css-1ftyaf0"
          >
            <span>
              Name
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <div
      className="MuiFormControl-root MuiTextField-root css-1cs2p1o-MuiFormControl-root-MuiTextField-root"
    >
      <label
        className="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-1pysi21-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink={false}
        htmlFor="Email"
        id="Email-label"
      >
        Email
      </label>
      <div
        className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          autoFocus={false}
          className="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
          disabled={false}
          id="Email"
          name="Email"
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder="Enter your email"
          required={false}
          type="text"
          value=""
        />
        <fieldset
          aria-hidden={true}
          className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            className="css-1ftyaf0"
          >
            <span>
              Email
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <button
      className="MuiLoadingButton-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root css-l07cfv-MuiButtonBase-root-MuiButton-root-MuiLoadingButton-root"
      disabled={false}
      id="Send"
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      Send
      <span
        className="MuiButton-endIcon MuiButton-iconSizeMedium css-9tj150-MuiButton-endIcon"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="SendIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M2.01 21 23 12 2.01 3 2 10l15 2-15 2z"
          />
        </svg>
      </span>
    </button>
  </div>
</section>
`;
